libsolidity/formal/SMTEncoder.cpp

```diff
@@ -1405,6 +1405,11 @@ bool SMTEncoder::visit(MemberAccess const& _memberAccess)
 					ContractDefinition const& contract = dynamic_cast<ContractType const&>(*magicType->typeArgument()).contractDefinition();
 					defineExpr(_memberAccess, contract.interfaceId());
 				}
+				else if (memberName == "typehash")
+				{
+					StructDefinition const& structDef = dynamic_cast<StructType const&>(*magicType->typeArgument()).structDefinition();
+					defineExpr(_memberAccess, u256(structDef.typehash()));
+				}
```

Comment:
@hrkrshnn @leonardoalt , this part only makes sure we can create an expression when we come across the `typehash` member access which will be roughly `(= X COMPUTED_HASH)`. So the tests in:
- https://github.com/ethereum/solidity/pull/14167/commits/ed3927b7f956a07fbfc400e0389edde8e0505bec

and the other ones I added (and then (re)moved) are basically testing whether these expressions can be created and be consumed by different smt engines that they would not `SAT`. So perhaps only one test is enough.

It's kind of the same with [`type_interfaceid.sol`](https://github.com/ethereum/solidity/blob/4216bb52003a97807be596245a68776ec35606f8/test/libsolidity/smtCheckerTests/types/type_interfaceid.sol) as well I guess.

Moved/created new semantic tests here:
- https://github.com/ethereum/solidity/pull/14167/commits/4383d05b6940346317f13c80a2575af6defa8349

I wasn't sure about how the diff test folders are arranged. Also, I think more semantic tests should be added as well as fuzzing and differential tests.